### ISSUE TITLE
feat: Attention Residuals (AttnRes) — learned depth-wise aggregation for transformer layers

### ISSUE THREAD

**Issue by dndungu** [zerfoo/zerfoo#103]
## Summary

Implement Attention Residuals (AttnRes) from [arXiv:2603.15031](https://arxiv.org/abs/2603.15031) (Kimi Team, March 2026). AttnRes replaces fixed additive residual connections with softmax attention over depth, allowing each layer to selectively aggregate previous layer outputs with learned, input-dependent weights.

This is a drop-in replacement for standard residual connections that yields a **1.25x compute advantage** at scale with <2% inference latency overhead.

## Problem

Standard residual connections accumulate all layer outputs with fixed unit weights:

```
h_l = h_{l-1} + f_{l-1}(h_{l-1})
```

This causes:
1. **Hidden-state growth**: magnitudes grow O(L) with depth, diluting each layer's contribution (PreNorm dilution)
2. **No selective access**: every layer receives the same uniformly-weighted sum — no mechanism to emphasize or suppress individual layer contributions
3. **Gradient imbalance**: earliest layers receive disproportionately large gradients

### Full AttnRes

```
h_l = sum(alpha_{i->l} * v_i)
alpha_{i->l} = softmax(w_l^T * RMSNorm(v_i))
```

Where `w_l` is a single learned d-dimensional pseudo-query per layer. O(L^2 * d) compute, O(L*d) memory.

### Block AttnRes (practical variant)

Partition L layers into N blocks (~8). Within each block: standard residual accumulation. Across blocks: softmax attention over N block-level representations.

- Memory: O(N*d) instead of O(L*d)
- Compute: O(N^2*d) instead of O(L^2*d)
- N=8 recovers most of the gain of full AttnRes

## Proposed API

```go
// In package layers/residual/

// AttnRes implements Attention Residuals — learned depth-wise aggregation.
// Each layer gets a pseudo-query w_l that attends over all previous layer outputs.
type AttnRes[T tensor.Numeric] struct {
    Query *graph.Parameter[T] // learned pseudo-query, shape [d]
    Norm  *normalization.RMSNorm[T]
}

// BlockAttnRes partitions layers into blocks with intra-block standard residuals
// and inter-block attention aggregation.
type BlockAttnRes[T tensor.Numeric] struct {
    BlockSize int
    Queries   []*graph.Parameter[T] // one per layer
    Norm      *normalization.RMSNorm[T]
}

// Forward computes the attended hidden state for the current layer.
// blocks: completed block representations [b0, ..., b_{n-1}]
// partialBlock: intra-block partial sum
// Returns: attended hidden state h_l
func (b *BlockAttnRes[T]) Forward(
    ctx context.Context,
    blocks []*tensor.TensorNumeric[T],
    partialBlock *tensor.TensorNumeric[T],
    layerIndex int,
) (*tensor.TensorNumeric[T], error)
```

## Integration Points

AttnRes is architecture-agnostic — it modifies how layers connect, not what layers do. Can be applied to:

1. **Standard transformers** (Llama, Gemma, Qwen, etc.) — replace additive residuals in transformer blocks
2. **HRM** — replace fixed residuals inside H-module and L-module transformer blocks
3. **Tabular MLPs** — replace residual connections in deep MLPs (ResNet tabular variant)

## Paper Results (48B MoE, 1.4T tokens)

| Benchmark | Baseline | AttnRes | Delta |
|-----------|----------|---------|-------|
| MMLU | 73.5 | 74.6 | +1.1 |
| GPQA-Diamond | 36.9 | 44.4 | +7.5 |
| Math | 53.5 | 57.1 | +3.6 |
| HumanEval | 59.1 | 62.2 | +3.1 |
| BBH | 76.3 | 78.0 | +1.7 |

Scaling law: Block AttnRes matches baseline trained with 1.25x more compute.

## Reference

- Paper: https://arxiv.org/abs/2603.15031
- Code: https://github.com/MoonshotAI/Attention-Residuals
- PyTorch pseudocode in paper Figure 2
